pkg/dss/cockroach/store.go

```
@@ -618,6 +626,159 @@ func (s *Store) insertIdentificationServiceAreaUnchecked(ctx context.Context, se
+				identification_service_areas
+			VALUES
+				($1, $2, $3, $4, $5, transaction_timestamp())
+			RETURNING
+				*`
+		upsertCellsForAreaQuery = `
+			UPSERT INTO
+				cells_identification_service_areas
+			VALUES
+				($1, $2, $3, transaction_timestamp())`
+		updateCellsForOtherAreaQuery = `
+			UPDATE
+				cells_identification_service_areas
+			SET
+				updated_at = transaction_timestamp()
+			WHERE
+				cell_id = ANY($1)`
+		deleteLeftOverCellsForAreaQuery = `
+			DELETE FROM
+				cells_identification_service_areas
+			WHERE
+				cell_id != ANY($1)
+			AND
+				identification_service_area_id = $2`
+		getAreaQuery = `
+			SELECT
+				COUNT(updated_at)
+			FROM
+				cells_identification_service_areas
+			WHERE
+				cell_id = ANY($1)
+			AND
+				updated_at != $2`
+		getAffectedSubscriberDetailsQuery = `
+			SELECT
+				subscriptions.id,
+				subscriptions.url,
+				subscriptions.notification_index
+			FROM
+				subscriptions
+			LEFT JOIN
+				(SELECT DISTINCT subscription_id FROM cells_subscriptions WHERE cell_id = ANY($1))
+			AS
+				affected_subscriptions
+			ON
+				affected_subscriptions.subscription_id = subscriptions.id
+			WHERE
+				subscriptions.owner != $2`
+	)
+
+	ts, err := versionStringToTimestamp(version)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	tx, err := s.Begin()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cids := make([]int64, len(cells))
+	for i, cell := range cells {
+		cids[i] = int64(cell)
+	}
+
+	versionMismatchCount := int64(0)
+	if err := tx.QueryRowContext(ctx, getAreaQuery, pq.Int64Array(cids), ts).Scan(&versionMismatchCount); err != nil {
+		return nil, nil, multierr.Combine(err, tx.Rollback())
+	}
+	if versionMismatchCount > 0 {
+		return nil, nil, multierr.Combine(
+			fmt.Errorf("version mismatch for identification service area %s", id),
+			tx.Rollback(),
+		)
+	}
+
+	isar := &identificationServiceAreaRow{}
+	if err := isar.scan(tx.QueryRowContext(ctx, upsertAreasQuery, id, owner, url, starts, ends)); err != nil {
```

for the insert case, we should return an error if it already exists. So we should probably separate this into 2 methods.

Also for the update we might also want to call a "get" first, since we'll have to return a different error code if it doesn't exist vs. if the version has changed.

---

Factored out into two functions and introduced a private `getIdentificationServiceArea` used by both functions. Also factored out `upsertIdentificationServiceArea` used by `{Create,Update}IdentificationServiceArea`.